### PR TITLE
Sharded Plugin 2/n: Allow ddp plugin to modify optimizer state saving

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Ties to #4178

Allows custom ddp plugin access to saving optimizer state. This is required for optimizer state sharding as we need to consolidate the optimizer state from all processes before saving.

cc @ananthsub @awaelchli 

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -202,6 +202,21 @@ def sync_tensor(self,
...
+    @property
+    def optimizer_states(self) -> List[dict]:
+        """
+        Returns states of optimizers. Allows for syncing/collating optimizer state from processes in custom
+        plugins.
+        Return:
+            Optimizer state dicts
+        """
+        if self.ddp_plugin:
+            return self.ddp_plugin.optimizer_states(self.trainer)
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/accelerator.py:214]: does this need to go through the DDP plugin? could the base accelerator define this as L215-218, and then the sharded ddp accelerator overrides `optimizer_states` more directly?

File: pytorch_lightning/accelerators/accelerator.py [hunk 2]
```diff
@@ -202,6 +202,21 @@ def sync_tensor(self,
...
+        optimizer_states = []
+        for i, optimizer in enumerate(self.trainer.optimizers):
+            optimizer_states.append(optimizer.state_dict())
+        return optimizer_states
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/accelerator.py:218]: this part is now also repeated both inside the accelerator and ddp plugin - i'm wondering how we can avoid the repetition

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -299,9 +299,14 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
...
-            optimizer_states = []
-            for i, optimizer in enumerate(self.trainer.optimizers):
-                optimizer_states.append(optimizer.state_dict())
+            if self.trainer.accelerator_backend:
+                # Rely on accelerator backend to obtain optimizer state from processes
+                optimizer_states = self.trainer.accelerator_backend.optimizer_states
+            else:
+                optimizer_states = []
+                for i, optimizer in enumerate(self.trainer.optimizers):
+                    optimizer_states.append(optimizer.state_dict())
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:308]: should this entire section rely on the trainer's accelerator backend for the optimizer states?

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py:214]: The sharded ddp accelerator is now a plugin (will update the original PR title), so all the sharded ddp logic lives within a plugin that is then put into each accelerator.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 3 above)
Comment (SeanNaren) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:308]: TIL we define a cpu accelerator, thought we don't... we should entirely rely on the backend, will modify :)

File: pytorch_lightning/accelerators/accelerator.py (hunk 2 above)
Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py]: I refactored the code to just override the state_dict getter, still dupes the `state_dict()` call but I think that's better than the loop of before!

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]
lgtm

**Review by williamFalcon — APPROVED** [review]
dope
